nixos/services.prometheus.exporters.ipmi: new module along with underlying ipmi_exporter package

snaar opened the pull request:
###### Description of changes

Bringing in new https://github.com/prometheus-community/ipmi_exporter exporter into existing Prometheus exporters framework.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

azahi commented on nixos/modules/services/monitoring/prometheus/exporters/ipmi.nix:
```diff
@@ -0,0 +1,54 @@
+        ${concatStringsSep " \\\n  " cfg.extraFlags}
+      '';
+
+      ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
```
Are you sure this works? I can't find any mention of this in the repo.

snaar replied:
I found it in code: https://github.com/prometheus-community/ipmi_exporter/blob/bb59b211ee3fe27d3327cc31a5441fa24bb6cb1b/main.go#L121

azahi replied:
Alright, I see it now. Thanks for clearing this up.

snaar commented on nixos/modules/services/monitoring/prometheus/exporters/ipmi.nix:
```diff
@@ -0,0 +1,54 @@
+  };
+
+  serviceOpts = let
+    checkedConfigFile = checkConfigLocation cfg.configFile;
+    checkedWebConfigFile = checkConfigLocation cfg.webConfigFile;
```
I'm not familiar with assertions. I've tried looking at examples of how it's used, but none of them are in similar context. Do you have any pointers on how it can be used in this kind of a module?

snaar replied:
Ok I think I figured it out, I'll move to assertions.

azahi replied:
Something along the lines of:
```
assertions = [
  {
    assertion = cfg.configFile != null && lib.hasPrefix "/tmp/" cfg.configFile;
    message = "Configuration file blah blah blah!";
  }
];
```

snaar replied:
Yeah, I just didn't realize it had to go into the prometheus/exporters.nix and not the ipmi.nix. Works good now.

snaar commented on nixos/modules/services/monitoring/prometheus/exporters/ipmi.nix:
```diff
@@ -0,0 +1,42 @@
+        (optionalString (cfg.webConfigFile != null) '' --web.config.file ${escapeShellArg cfg.webConfigFile} '') +
+        (optionalString (cfg.configFile != null) '' --config.file ${escapeShellArg cfg.configFile} '') + '' \
+        ${concatStringsSep " \\\n  " cfg.extraFlags}
+      '';
```
Yeah this is so much nicer, thanks. Someone should go through other prometheus exporters and clean them up - I was using them as examples.

azahi commented:
Result of `nixpkgs-review pr 188754` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>prometheus-ipmi-exporter</li>
  </ul>
</details>

snaar commented:
Is that result unusual/unexpected? I can't tell.

azahi commented:
> Is that result unusual/unexpected? I can't tell.

I'm not sure. I've tested stuff locally and it works fine. Since all ofborg jobs pass, I guess everything is OK.

azahi reviewed (commented):
Could you split your commit in two: one which adds a package and one which adds a module. This will also allow you to make the commit message a bit shorter.

snaar commented:
Sure. Keeping it in one PR okay?

azahi commented:
> Sure. Keeping it in one PR okay?

Yes!

snaar committed:
ipmi_exporter: init at 1.6.1

snaar committed:
nixos/services.prometheus.exporters.ipmi: new module

Bringing in new https://github.com/prometheus-community/ipmi_exporter exporter into existing Prometheus exporters framework.

azahi reviewed (changes requested):
The module looks good! I only have a few suggestions regarding the package itself now.

snaar commented:
Ugh, rebase went back too far...

snaar commented:
I don't actually know how to fix this, I should probably recreate PR.

azahi commented:
You can just force push to a new branch with the **same** name. There's no need to create a new PR.

snaar commented:
I had to rebase then delete few commits, since previous rebase "touched" 9 extra commits making them diverge from master from what I understand. So that worked.

snaar commented:
(Since because I split commit into two, I had to edit the "middle" commit, and git UX is atrocious.)

IvarWithoutBones commented on nixos/modules/services/monitoring/prometheus/exporters/ipmi.nix:
```diff
@@ -0,0 +1,41 @@
+  logPrefix = "services.prometheus.exporter.ipmi";
+  cfg = config.services.prometheus.exporters.ipmi;
+in {
+  port = 9290;
```
Shouldn't this be an configurable option?

azahi replied:
It's handled in the parent module [here](https://github.com/NixOS/nixpkgs/blob/1f709c16104f43d0e67a14a5243ae94f1e03bcaf/nixos/modules/services/monitoring/prometheus/exporters.nix#L11).

IvarWithoutBones reviewed (approved):
LGTM :+1: 

Result of `nixpkgs-review pr 188754` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>prometheus-ipmi-exporter</li>
  </ul>
</details>
